internal/pkg/cli/deploy/deploy.go

```diff
@@ -0,0 +1,873 @@
+}
+
+// ServiceDeployer uses CloudFormation to deploy a service.
+type ServiceDeployer interface {
+	DeployService(out progress.FileWriter, conf cloudformation.StackConfiguration, bucketName string, opts ...awscloudformation.StackOption) error
+}
+
+// ServiceForceUpdater force updates a service.
+type ServiceForceUpdater interface {
+	ForceUpdateService(app, env, svc string) error
+	LastUpdatedAt(app, env, svc string) (time.Time, error)
+}
+
+// WorkspaceReader reads from the workspace.
+type WorkspaceReader interface {
+	manifestReader
+	pathFinder
+}
+
+// Interpolator interpolates a manifest.
+type Interpolator interface {
+	Interpolate(s string) (string, error)
+}
+
+// progress is the interface to inform the user that a long operation is taking place.
+type Progress interface {
+	Start(label string)
+	Stop(label string)
+}
+
+type manifestReader interface {
+	ReadWorkloadManifest(name string) (workspace.WorkloadManifest, error)
+}
+
+type pathFinder interface {
+	Path() (string, error)
+}
+
+// WorkloadDeployer holds all the metadata and clients needed to deploy a workload.
+type WorkloadDeployer struct {
+	Name     string
+	App      *config.Application
+	Env      *config.Environment
+	ImageTag string
+	S3Bucket string
+
+	// cached variables
+	mft           interface{}
+	workspacePath string
+}
+
+// UploadArtifactsInput is the input of UploadArtifacts.
+type UploadArtifactsInput struct {
+	ShouldUploadArtifacts bool
+
+	FS                 *afero.Afero
+	Uploader           Uploader
+	Templater          Templater
+	WS                 WorkspaceReader
+	NewInterpolator    func(app, env string) Interpolator
+	Unmarshal          func([]byte) (manifest.WorkloadManifest, error)
+	ImageBuilderPusher ImageBuilderPusher
+}
+
+// UploadArtifactsOutput is the output of UploadArtifacts.
+type UploadArtifactsOutput struct {
+	ImageDigest string
+	EnvFileARN  string
+	AddonsURL   string
+}
+
+// DeployWorkloadInput is the input of DeployWorkload.
+type DeployWorkloadInput struct {
+	ImageDigest    string
+	EnvFileARN     string
+	AddonsURL      string
+	RootUserARN    string
+	AppTags        map[string]string
+	ResourceTags   map[string]string
+	ImageRepoURL   string
+	ForceNewUpdate bool
+
+	WS                     WorkspaceReader
+	NewInterpolator        func(app, env string) Interpolator
+	Unmarshal              func([]byte) (manifest.WorkloadManifest, error)
+	UploadOpts             *UploadCustomResourcesOpts
+	SNSTopicsLister        SNSTopicsLister
+	ServiceDeployer        ServiceDeployer
+	NewSvcUpdater          func(func(*session.Session) ServiceForceUpdater)
+	NewAppVersionGetter    func(string) (VersionGetter, error)
+	PublicCIDRBlocksGetter PublicCIDRBlocksGetter
+	ServiceForceUpdater    ServiceForceUpdater
+	EndpointGetter         EndpointGetter
+	Spinner                Progress
+
+	now func() time.Time
+}
+
+// DeployWorkloadOutput is the output of DeployWorkload.
+type DeployWorkloadOutput struct {
+	RDWSAlias     string
```

Comment:
Hmm I did this because if you look at aws sdk, the output of their APIs leave some fields empty if they are not applicable. For example, if the service is Fargate then some EC2 fields are not set in the output. Also, I thought it would be easier to use as well. What advantage do you think over the current implementation 🤔

Reply:
I'm not strongly opposed to it given that it's just one field at the moment, so if you want to feel free to proceed.  

But I don't agree that it's easier to use 🤔  How will this type progress in the future? My guess is like this:
```go
type DeployWorkloadOutput struct {
	RDWSAlias     string
	LBWSAlias     string
	BackendAlias string
	Subscriptions []manifest.TopicSubscription
}
```
Now each one of these fields need to be individually checked for their zero-value by the client, the `DeployWorkloadOutput` is a mixed bag of mutually exclusive fields. Instead compare it to using duck-typing: https://gist.github.com/efekarakus/6dd25c0dfc11badf5d506dd177bf881d

Reply:
Ohh gotcha. I think it's because i haven't put it into the real production yet. I'll make this change when integrating this new package and removing the old codes.

Reply:
Seems like it is not easy to do what you proposed above because the order of recommendation becomes a problem :( let me know your thought on [this PR](https://github.com/aws/copilot-cli/pull/3239)!

Reply:
Nvm boss updated the new PR to address this fb